### ISSUE TITLE
Implement `starknet_getEvents` RPC method

### ISSUE THREAD

**Issue by FabijanC** [Shard-Labs/starknet-devnet#231]
Was not handled by #220:

https://github.com/Shard-Labs/starknet-devnet/pull/220#issuecomment-1220776804:
> this is quite a big method, which will mutate a bit in the future, but definetly is a candidate for an issue to be implemented in

**Comment by mikiw** [comment]
Specification https://github.com/starkware-libs/starknet-specs/blob/v0.1.0/api/starknet_api_openrpc.json

**Comment by mikiw** [comment]
I'm just reading slack and there will be different filtering logic compared to specification.

**Comment by mikiw** [comment]
Here is the second version of the spec:
https://github.com/starkware-libs/starknet-specs/blob/v0.2.0/api/starknet_api_openrpc.json

Implementation of the filters can be found here:
https://github.com/eqlabs/pathfinder

**Comment by mikiw** [comment]
Current filtering logic in [pathfinder](https://github.com/eqlabs/pathfinder)

```
                let keys = row.get_ref_unwrap("keys").as_str().unwrap();

                // no need to allocate a vec for this in loop
                let mut temp = [0u8; 32];

                let keys: Vec<_> = keys
                    .split(' ')
                    .map(|key| {
                        let used =
                            base64::decode_config_slice(key, base64::STANDARD, &mut temp).unwrap();
                        let key = StarkHash::from_be_slice(&temp[..used]).unwrap();
                        EventKey(key)
                    })
                    .collect();
```

**Comment by mikiw** [comment]
Why invokes generate estimation fee events? Is that expected behaviour?

**Comment by FabijanC** [comment]
In tests we are using Starknet CLI which performs fee estimation before invocation.

When you say events, do you mean like actual starknet events?

`RESULT_PAGE_REQUEST` is defined here, is that of any help? https://github.com/starkware-libs/starknet-specs/blob/v0.1.0/api/starknet_api_openrpc.json#L658

**Comment by FabijanC** [comment]
Ok, the estimation fee event (which I didn't even know exists), probably comes from the fact that Starknet CLI (used in our tests), performs fee estimation if no max fee is provided.

Regarding which version should be implemented, what is the difference between getEvents in 0.1.0 and 0.2.0?

**Comment by mikiw** [comment]
There is a difference in paging, in v0.2.0 there is a `chunk_size` and `continuation_token`.

**Comment by FabijanC** [comment]
Does 0.2.0 impose breaking changes? Are getEvents requests from 0.1.0 a subset of those in 0.2.0 and thus acceptable?

**Comment by mikiw** [comment]
These are not breaking changes and we are already compatible with 0.2.0 we use `chunk_size` and `continuation_token`. Just `continuation_token` is a string with next natural numbers like 1, 2, 3 etc.

**Comment by FabijanC** [comment]
Ok then, if it's not a breaking change, this feature of 0.2.0 can already be merged before the 0.2.0 PR itself.
